data-contract/src/contract/invalidate.rs

```diff
@@ -28,27 +36,90 @@ impl CertificationContract {
 
         let lookup = self.tokens.token_metadata_by_id.as_mut().unwrap();
 
-        let metadata = lookup
-            .get(&token_id)
-            .expect("Token does not exist");
+        let metadata = lookup.get(&token_id).expect("Token does not exist");
 
-        let certification_metadata = serde_json::from_str::<CertificationExtraMetadata>(&metadata.extra.unwrap()).unwrap();
+        let certification_metadata =
+            serde_json::from_str::<CertificationExtraMetadata>(&metadata.extra.unwrap()).unwrap();
 
         let recipient_id = certification_metadata.original_recipient_id.clone();
 
-        lookup
-            .insert(&token_id, &TokenMetadata {
-                extra: Some(CertificationExtraMetadata {
-                    valid: false,
-                    ..certification_metadata
-                }.to_json()),
+        lookup.insert(
+            &token_id,
+            &TokenMetadata {
+                extra: Some(
+                    CertificationExtraMetadata {
+                        valid: false,
+                        ..certification_metadata
+                    }
+                    .to_json(),
+                ),
                 ..metadata
-            });
+            },
+        );
 
         self.create_event_log(CertificationEventLogData::Invalidate {
-            token_id,
+            token_id: token_id.to_owned(),
             recipient_id,
             memo,
-        }).emit();
+        })
+        .emit();
+
+        if let Some(ref trash_account) = self.trash_account.get() {
+            let owner_id = self.tokens.owner_by_id.get(&token_id).unwrap();
+            self.tokens
+                .internal_transfer(&owner_id, trash_account, &token_id, None, None);
+        }
```

Comment:
@ryancwalsh This is the new code relevant to certification _invalidation_ (and trash accounts).